client/beefy/src/worker.rs

```diff
@@ -92,17 +92,55 @@ pub(crate) struct VoterOracle<B: Block> {
 }
 
 impl<B: Block> VoterOracle<B> {
-	pub fn genesis(
-		session_start: NumberFor<B>,
-		validator_set: ValidatorSet<AuthorityId>,
+	/// Verify provided `sessions` satisfies requirements, then build `VoterOracle`.
+	pub fn checked_new(
+		sessions: VecDeque<Rounds<Payload, B>>,
 		min_block_delta: u32,
-	) -> Self {
-		let mut sessions = VecDeque::new();
-		sessions.push_back(Rounds::new(session_start, validator_set));
-		Self {
-			sessions,
-			// Always target at least one block better than current best beefy.
-			min_block_delta: min_block_delta.max(1),
+	) -> Option<Self> {
+		let mut prev_start = Zero::zero();
+		let mut prev_validator_id = None;
+		// verifies the
+		let mut validate = || -> bool {
+			if sessions.is_empty() {
+				println!("🥩 empty sessions");
```

Comment:
Probably return `Result<(), String>` from this closure && use trace outside to print the error (instead of println)?

Reply:
oh, the `println` are leftover from some test debug, I need to remove them. good catch!